en/ros2/px4_sdk.md

```diff
@@ -0,0 +1,336 @@
+
+- A mode executor is an optional component one level higher than a mode. It is a state machine that can activate modes, and wait for their completion.
+- It can only do so while it is in charge.
+  For that, an executor has exactly one _owned mode_ (and a mode can be owned by at most one executor).
+  This mode serves as activation for the executor: when the user selects the mode, the owning executor gets activated and can select any mode.
+  It stays in charge until the user switches modes (by RC or from a GCS), or a failsafe triggers a mode switch.
+  Should the failsafe clear, the executor gets reactivated.
+- This allows multiple executors to coexist.
+- Executors cannot activate other executors.
+- Within the SDK, a mode executor is always implemented in combination with a custom mode.
+
+:::note
+- These definitions guarantee that a user can take away control from a custom mode or executor at any point in time by commanding a mode switch through RC or a GCS.
+- A mode executor is transparent to the user.
+  It gets indirectly selected and activated through the owning mode, and thus the mode should be named accordingly.
+:::
+
+##### When would you use a mode executor?
+
+The most common use-case is the execution of an autonomous 'mission', such as a survey or a payload delivery: the executor triggers a take-off, then switches to a custom mode which performs the survey, and when completed, triggers an RTL.
+
+#### Configuration Overrides
+
+Both modes and executors can define configuration overrides, allowing to customize certain behaviors while the mode or executor is active.
+
+These are currently implemented:
+
+- Disabling auto-disarm. This allows for landing and then taking off again (e.g. to release a payload).
+- Ability to defer non-essential failsafes.
+  It allows for example an executor to run a critical action without being interrupted by low-battery failsafe, such as controlling a winch.
+
+### Comparison to Offboard Control
+
+The above concepts provide a number of advantages over traditional [offboard control](../ros/offboard_control.md):
+
+- Multiple nodes or applications can coexist and even run at the same time. It is clearly defined which node can control the vehicle at any given time.
+- Exposing a selectable mode to the GCS.
+- Integrated with the failsafe state machine and arming checks.
+- It is well-defined which setpoint types can be sent.
+- It is possible to replace an FMU-internal mode (such as RTL).
+
+## Getting Started
+
+The following steps are required to get started:
+
+1. Make sure you have a working [ROS 2 setup](../ros/ros2_comm.md), with _px4_msgs_ in the ROS 2 workspace.
+2. Clone the SDK into the workspace:
+
+   ```shell
+   cd $ros_workspace/src
+   git clone --recursive https://github.com/PX4/px4_sdk.git
+   ```
+
+   :::note
+   To ensure compatibility, use the latest _main_ branches for PX4, _px4_msgs_ and the SDK.
+   See also [here](https://github.com/PX4/px4_sdk#compatibility-with-px4).
```

Comment:
This makes me uncomfortable - it basically says we ensure compatibility of the SDK by forcing you to use an unstable version of PX4. The linked doc is better in that it says we might change this in future. We should come up with a plan now and let people know.

The linked doc mentions this, which should also appear in this doc (I haven't checked) but it is cool:

> The SDK checks for message compatibility on startup when registering a mode. ALL_PX4_SDK_MESSAGES defines the set of checked messages. If you use other messages, you can check them using:
```
if (!px4_sdk::messageCompatibilityCheck(node, {{"/fmu/in/vehicle_rates_setpoint"}})) {
  throw std::runtime_error("Messages incompatible");
}
```